Simplify SVGTextQuery::modifyStartEndPositionsRespectingLigatures

Ahmad-S792 opened the pull request:
#### 006900da2e50db3ee2c2cc1b490de3b0a2ab6721
<pre>
Simplify SVGTextQuery::modifyStartEndPositionsRespectingLigatures

<a href="https://bugs.webkit.org/show_bug.cgi?id=262458">https://bugs.webkit.org/show_bug.cgi?id=262458</a>

Reviewed by Chris Dumez.

Merge: <a href="https://src.chromium.org/viewvc/blink?view=revision&amp">https://src.chromium.org/viewvc/blink?view=revision&amp</a>;revision=177231

Simplify this method by passing it the fragment in question, and then use
SVGTextFragment::metricsListOffset to get a reasonable starting point.
Rewrite the big loop into two simpler loops - one for each end-point.
This shaves another ~15% off the TC in the bug (on top of 268656@main).

* Source/WebCore/rendering/svg/SVGTextQuery.cpp:
(SVGTextQuery::mapStartEndPositionsIntoFragmentCoordinates):
(SVGTextQuery::modifyStartEndPositionsRespectingLigatures):
* Source/WebCore/rendering/svg/SVGTextQuery.h:

Canonical link: <a href="https://commits.webkit.org/268751@main">https://commits.webkit.org/268751@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/965966790e575c173a0387d457847007605d0525

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20516 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/20941 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21587 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22413 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19141 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/20748 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24198 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21117 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/20535 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20735 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20596 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17836 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23269 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17764 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18641 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/24926 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18841 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18817 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22865 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19411 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16489 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18624 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4935 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/22964 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19230 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/rendering/svg/SVGTextQuery.cpp:
```diff
@@ -155,82 +155,45 @@ bool SVGTextQuery::mapStartEndPositionsIntoFragmentCoordinates(Data* queryData,
-        lastPositionOffset = positionOffset;
-        positionOffset += metrics.length();
+        fragmentOffsetInBox = glyphEnd;
+        textMetricsOffset++;
```
nit: `++textMetricsOffset;`

Ahmad-S792 replied:
Done in [9659667](https://github.com/WebKit/WebKit/commit/965966790e575c173a0387d457847007605d0525).

cdumez commented on Source/WebCore/rendering/svg/SVGTextQuery.cpp:
```diff
@@ -155,82 +155,45 @@ bool SVGTextQuery::mapStartEndPositionsIntoFragmentCoordinates(Data* queryData,
-            endPosition = positionOffset;
+    // Find the text metrics cell that contain or ends at the character endPosition.
+    while (fragmentOffsetInBox < fragmentEndInBox) {
+        SVGTextMetrics& metrics = textMetricsValues[textMetricsOffset];
```
nit: `auto&`

Ahmad-S792 replied:
Done in [9659667](https://github.com/WebKit/WebKit/commit/965966790e575c173a0387d457847007605d0525).

cdumez commented on Source/WebCore/rendering/svg/SVGTextQuery.cpp:
```diff
@@ -155,82 +155,45 @@ bool SVGTextQuery::mapStartEndPositionsIntoFragmentCoordinates(Data* queryData,
+        fragmentOffsetInBox += metrics.length();
+        if (fragmentOffsetInBox >= endPosition)
+            break;
+        textMetricsOffset++;
```
nit: `++textMetricsOffset`

Ahmad-S792 replied:
Done in [9659667](https://github.com/WebKit/WebKit/commit/965966790e575c173a0387d457847007605d0525).

Ahmad-S792 committed:
Simplify SVGTextQuery::modifyStartEndPositionsRespectingLigatures

https://bugs.webkit.org/show_bug.cgi?id=262458

Reviewed by Chris Dumez.

Merge: https://src.chromium.org/viewvc/blink?view=revision&revision=177231

Simplify this method by passing it the fragment in question, and then use
SVGTextFragment::metricsListOffset to get a reasonable starting point.
Rewrite the big loop into two simpler loops - one for each end-point.
This shaves another ~15% off the TC in the bug (on top of 268656@main).

* Source/WebCore/rendering/svg/SVGTextQuery.cpp:
(SVGTextQuery::mapStartEndPositionsIntoFragmentCoordinates):
(SVGTextQuery::modifyStartEndPositionsRespectingLigatures):
* Source/WebCore/rendering/svg/SVGTextQuery.h:

Canonical link: https://commits.webkit.org/268751@main
